File: typescript/src/surfacing.ts
import type { AffectState } from './contracts.js';
import { createAffectState } from './contracts.js';
import { pyRound } from './internal/round.js';
import { rstripPyWhitespace, splitPyWhitespace, stripPyWhitespace } from './internal/whitespace.js';

function smartTrim(text: string, maxChars: number): string {
  const cleaned = splitPyWhitespace(text).join(' ');
  if (maxChars <= 0 || cleaned.length <= maxChars) return cleaned;
  const window = cleaned.slice(0, maxChars);
  const sentenceEnd = Math.max(window.lastIndexOf('.'), window.lastIndexOf('!'), window.lastIndexOf('?'));
  if (sentenceEnd >= maxChars * 0.5) return rstripPyWhitespace(window.slice(0, sentenceEnd + 1));
  const lastSpace = window.lastIndexOf(' ');
  if (lastSpace === -1) return rstripPyWhitespace(window);
  const trimmed = rstripPyWhitespace(window.slice(0, lastSpace)).replace(/[,;:—-]+$/u, '');
  return trimmed ? `${trimmed}…` : rstripPyWhitespace(window);
}

export function cleanText(value: unknown, maxLength: number): string | null {
  if (typeof value !== 'string') return null;
  const unquoted = stripPyWhitespace(value).replace(/^"+|"+$/g, '');
  const cleaned = splitPyWhitespace(unquoted).join(' ');
  return cleaned ? smartTrim(cleaned, maxLength) : null;
}

export interface PresenceLineInput {
  reflectionPending: boolean;
  dueIntention: boolean;
  carriedThought: boolean;
  openThread: boolean;
}

/**
 * The rendered wording of the presence surface.
 *
 * The defaults are the reference conversational phrasing. They are English and
 * companion-shaped on purpose; a game, tutoring, or support domain should pass
 * its own labels rather than surfacing "here with you" to its users.
 */
export interface PresenceLabels {
  readonly reflecting: string;
  readonly dueIntention: string;
  readonly carriedThought: string;
  readonly openThread: string;
  readonly idle: string;
  readonly reflectingMode: string;
  readonly idleMode: string;
}

export const DEFAULT_PRESENCE_LABELS: PresenceLabels = Object.freeze({
  reflecting: 'still reflecting',
  dueIntention: 'holding a follow-up',
  carriedThought: 'carrying a thread',
  openThread: 'holding a pattern',
  idle: 'here with you',
  reflectingMode: 'reflecting',
  idleMode: 'quiet',
});

export function presenceLine(
  input: PresenceLineInput,
  labels: PresenceLabels = DEFAULT_PRESENCE_LABELS,
): string {
  if (input.reflectionPending) return labels.reflecting;
  if (input.dueIntention) return labels.dueIntention;
  if (input.carriedThought) return labels.carriedThought;
  if (input.openThread) return labels.openThread;
  return labels.idle;
}

export interface CognitionState {
  reflectionPending?: boolean;
  carriedThought?: boolean;
  dueIntention?: boolean;
  openThread?: boolean;
  intentionality?: boolean;
  curiosity?: boolean;
}

export interface PresenceVector {
  trust: number;
  valence: number;
  arousal: number;
  longing: number;
  awaiting: boolean;
  mode: string;
  line: string;
  affect: { valence: number; arousal: number; dominance: number };
  relationship: { stage: string; trust: number; longing: number; comfort: number };
  cognition: {
    reflection_pending: boolean;
    carried_thought: boolean;
    due_intention: boolean;
    open_thread: boolean;
    intentionality: boolean;
    curiosity: boolean;
  };
  source: 'affect_state';
}

export function buildPresenceVector(
  stateInput: AffectState,
  cognitionInput: CognitionState = {},
  labels: PresenceLabels = DEFAULT_PRESENCE_LABELS,
): PresenceVector {
  const state = createAffectState(stateInput);
  const cognition: Required<CognitionState> = {
    reflectionPending: cognitionInput.reflectionPending ?? false,
    carriedThought: Boolean(cleanText(state.carriedThought, 300)),
    dueIntention: cognitionInput.dueIntention ?? false,
    openThread: cognitionInput.openThread ?? false,
    intentionality: cognitionInput.intentionality ?? false,
    curiosity: cognitionInput.curiosity ?? false,
  };
  const trust = pyRound(state.relationship.trustScore, 4);
  const valence = pyRound(state.mood.valence, 4);
  const arousal = pyRound(state.mood.arousal, 4);
  const longing = pyRound(state.attachment.longing, 4);
  return {
    trust,
    valence,
    arousal,
    longing,
    awaiting: cognition.reflectionPending,
    mode: cognition.reflectionPending ? labels.reflectingMode : labels.idleMode,
    line: presenceLine(cognition, labels),
    affect: { valence, arousal, dominance: pyRound(state.mood.dominance, 4) },
    relationship: {
      stage: state.relationship.stage,
      trust,
      longing,
      comfort: pyRound(state.attachment.comfort, 4),
    },
    cognition: {
      reflection_pending: cognition.reflectionPending,
      carried_thought: cognition.carriedThought,
      due_intention: cognition.dueIntention,
      open_thread: cognition.openThread,
      intentionality: cognition.intentionality,
      curiosity: cognition.curiosity,
    },
    source: 'affect_state',
  };
}

File: python/src/affect_kernel/surfacing.py
"""Pure state-to-presence transforms with all external cognition injected."""

from __future__ import annotations

from dataclasses import dataclass, replace

from .models import (
    AffectState,
    CognitionState,
    PresenceAffect,
    PresenceRelationship,
    PresenceVector,
)


def _smart_trim(text: object, max_chars: int) -> str:
    cleaned = " ".join(str(text or "").strip().split())
    if max_chars <= 0 or len(cleaned) <= max_chars:
        return cleaned
    window = cleaned[:max_chars]
    sentence_end = max(window.rfind(mark) for mark in (".", "!", "?"))
    if sentence_end >= max_chars * 0.5:
        return window[: sentence_end + 1].rstrip()
    last_space = window.rfind(" ")
    if last_space == -1:
        return window.rstrip()
    trimmed = window[:last_space].rstrip().rstrip(",;:—-")
    return f"{trimmed}…" if trimmed else window.rstrip()


def clean_text(value: object, max_len: int) -> str | None:
    """Normalize a possible surfaced string, then shorten it at a word boundary."""
    if not isinstance(value, str):
        return None
    cleaned = " ".join(value.strip().strip('"').split())
    if not cleaned:
        return None
    return _smart_trim(cleaned, max_len)


@dataclass(frozen=True, slots=True)
class PresenceLabels:
    """The rendered wording of the presence surface.

    The defaults are the reference conversational phrasing. They are English and
    companion-shaped on purpose; a game, tutoring, or support domain should pass
    its own labels rather than surfacing "here with you" to its users.
    """

    reflecting: str = "still reflecting"
    due_intention: str = "holding a follow-up"
    carried_thought: str = "carrying a thread"
    open_thread: str = "holding a pattern"
    idle: str = "here with you"
    reflecting_mode: str = "reflecting"
    idle_mode: str = "quiet"

    def __post_init__(self) -> None:
        for name in (
            "reflecting",
            "due_intention",
            "carried_thought",
            "open_thread",
            "idle",
            "reflecting_mode",
            "idle_mode",
        ):
            value = getattr(self, name)
            if not isinstance(value, str):
                raise TypeError(f"{name} must be a string")
            if not value.strip():
                raise ValueError(f"{name} must not be empty")


DEFAULT_PRESENCE_LABELS = PresenceLabels()


def presence_line(
    *,
    reflection_pending: bool,
    due_intention: bool,
    carried_thought: bool,
    open_thread: bool,
    labels: PresenceLabels | None = None,
) -> str:
    """Select one compact presence label using a fixed priority cascade."""
    chosen = labels or DEFAULT_PRESENCE_LABELS
    if reflection_pending:
        return chosen.reflecting
    if due_intention:
        return chosen.due_intention
    if carried_thought:
        return chosen.carried_thought
    if open_thread:
        return chosen.open_thread
    return chosen.idle


def build_presence_vector(
    state: AffectState,
    cognition: CognitionState | None = None,
    *,
    labels: PresenceLabels | None = None,
) -> PresenceVector:
    """Build a UI-ready presence snapshot without reading persistence or services."""
    chosen = labels or DEFAULT_PRESENCE_LABELS
    supplied = cognition or CognitionState()
    carried = bool(clean_text(state.carried_thought, 300))
    resolved_cognition = replace(supplied, carried_thought=carried)
    line = presence_line(
        reflection_pending=resolved_cognition.reflection_pending,
        due_intention=resolved_cognition.due_intention,
        carried_thought=resolved_cognition.carried_thought,
        open_thread=resolved_cognition.open_thread,
        labels=chosen,
    )
    trust = round(state.relationship.trust, 4)
    valence = round(state.mood.valence, 4)
    arousal = round(state.mood.arousal, 4)
    longing = round(state.attachment.longing, 4)
    return PresenceVector(
        trust=trust,
        valence=valence,
        arousal=arousal,
        longing=longing,
        awaiting=resolved_cognition.reflection_pending,
        mode=chosen.reflecting_mode if resolved_cognition.reflection_pending else chosen.idle_mode,
        line=line,
        affect=PresenceAffect(
            valence=valence,
            arousal=arousal,
            dominance=round(state.mood.dominance, 4),
        ),
        relationship=PresenceRelationship(
            stage=state.relationship.stage,
            trust=trust,
            longing=longing,
            comfort=round(state.attachment.comfort, 4),
        ),
        cognition=resolved_cognition,
    )


__all__ = [
    "DEFAULT_PRESENCE_LABELS",
    "PresenceLabels",
    "build_presence_vector",
    "clean_text",
    "presence_line",
]
